### ISSUE TITLE
Publish immutable, checksum-verified model bundles and pin Modal deployments

### ISSUE THREAD

**Issue by TumeloKonaite** [TumeloKonaite/Customer-Churning-Repo#95]
## Summary

Extend the existing DagsHub MLflow model workflow with a versioned integrity manifest that verifies the complete set of production artifacts belonging to a registered model version.

DagsHub MLflow remains the only tracking backend and model registry. This issue must not introduce a second filesystem registry, content-addressed model store, custom `models/{hash}` hierarchy, or alternative model-version identity.

The existing numeric DagsHub registered-model version remains authoritative:

```text
models:/churn_predictor/<numeric-version>
```

The existing application identity remains:

```text
dagshub:<owner>/<repository>:churn_predictor:<numeric-version>
```

The purpose of this issue is narrower: independently verify that the complete collection of artifacts associated with that registered version has not been removed, replaced, or corrupted.

## Motivation

The current implementation already verifies the fitted MLflow model and its pipeline checksum. It also validates contracts, references, metrics, source lineage, model signature, input example, and model identity.

However, the pipeline checksum alone does not cryptographically cover every production-relevant run artifact. For example, an individual contract, reference dataset, evaluation file, or dependency file could be changed without changing the existing pipeline checksum.

A production model version should therefore have one additional integrity identity:

```text
artifact_manifest_sha256
```

This digest will cover a canonical manifest containing the paths, roles, sizes, and SHA-256 digests of every protected artifact.

This integrity checksum supplements the DagsHub model version. It does not replace it.

## Existing functionality not included in this issue

The following capabilities are already implemented and must not be recreated:

- Local training through `DataIngestion`, `ModelTrainer`, and `TrainingPipeline`.
- Multi-model comparison and validation-based selection.
- Logging the exact fitted pipeline to DagsHub MLflow.
- Explicit MLflow model signature and input example.
- Numeric `churn_predictor` registry versions.
- Source revision, dataset identity, metrics, parameters, contracts, and reference logging.
- Separate drift and evaluation references.
- Prohibited-identifier and privacy validation.
- Exact-version registered-model validation.
- Inference-only `build/model` packaging.
- Exact Modal model/run/version/checksum pinning.
- Modal startup validation.
- Verified model identity in health and prediction responses.
- Neon connectivity and Alembic ownership.

Do not introduce parallel implementations of these capabilities.

## Objective

For every newly registered production-eligible `churn_predictor` version:

1. Generate a canonical integrity manifest covering all required production artifacts.
2. Calculate a SHA-256 digest of that manifest.
3. Record the digest on the DagsHub MLflow run and registered model version.
4. Publish a completion marker only after registration and integrity validation succeed.
5. Extend the existing validator to independently recalculate and verify the manifest.
6. Pin Modal deployment preparation to the expected manifest checksum.
7. Expose the verified manifest checksum through health metadata and safe operational logs.

## DagsHub-only architecture

DagsHub MLflow must remain the single remote source of truth.

Do not add:

- `models/{model_name}/{content_hash}/` as another registry;
- a local historical model store;
- a bucket-storage abstraction;
- a `latest` model directory or pointer;
- a second model publication CLI;
- a content hash that replaces the numeric MLflow model version;
- training or model selection inside Modal; or
- references and evaluation rows inside the Modal image.

The existing `src.mlops` module and `validate-model` command must be extended instead of introducing a parallel artifact subsystem.

## Required integrity artifacts

Each newly completed registered-model run must contain:

```text
integrity/
├── artifact_manifest.json
└── publication_complete.json
```

A versioned JSON Schema for `artifact_manifest.json` must be stored in the repository under an appropriate contracts directory.

The manifest and completion marker must not contain credentials, connection strings, authentication headers, raw customer identifiers, or raw feature rows.

## Artifact manifest

`integrity/artifact_manifest.json` must include at least:

- manifest schema version;
- canonicalization version;
- DagsHub repository owner and repository name;
- MLflow experiment name;
- MLflow run ID;
- registered model name;
- exact numeric registered model version;
- canonical `model_version_id`;
- source commit SHA;
- source branch or release identifier;
- dirty-worktree status;
- selected classifier name;
- selection metric;
- feature-schema version;
- prediction-contract version;
- prediction-horizon contract version;
- identity/privacy contract version;
- monitoring-contract version;
- training dataset identity;
- classification threshold;
- positive-class definition;
- Python version;
- scikit-learn version;
- MLflow version;
- pipeline tree checksum;
- protected artifact entries; and
- publication timestamp in UTC.

Each protected artifact entry must contain:

```json
{
  "path": "contracts/feature_schema.json",
  "role": "feature_contract",
  "sha256": "<64-character digest>",
  "size_bytes": 1234
}
```

Artifact paths must be normalized repository-independent POSIX paths.

The manifest must not contain absolute local filesystem paths.

## Protected artifact set

The manifest must cover, at minimum:

- the complete logged MLflow model directory;
- `MLmodel`;
- serialized fitted pipeline;
- MLflow dependency and environment files;
- model input example;
- serving input example, when present;
- feature schema;
- prediction contract;
- privacy contract;
- monitoring contract;
- training configuration;
- evaluation metrics;
- model-comparison results;
- dataset identity metadata;
- drift-reference metadata;
- drift-reference dataset;
- evaluation-reference metadata; and
- evaluation-reference dataset.

The implementation may include additional approved artifacts.

Transient files, logs, credentials, `.env` files, cache files, unrestricted source datasets, and local absolute paths must never be included.

## Pipeline checksum

Retain the existing pipeline checksum for compatibility.

The manifest must record:

```text
pipeline_sha256
```

The validator must confirm that this value matches:

1. the registered-model version tag;
2. the MLflow run tag or parameter;
3. the deployment expectation, when supplied; and
4. the independently recalculated checksum of the downloaded MLflow model.

The new manifest checksum must not replace the pipeline checksum. They protect different scopes:

- `pipeline_sha256` protects the logged inference model.
- `artifact_manifest_sha256` protects the canonical artifact inventory and its digests.

## Canonical serialization

Define one deterministic JSON serialization rule for the manifest:

- UTF-8 encoding;
- sorted object keys;
- no insignificant whitespace;
- stable list ordering;
- normalized POSIX artifact paths;
- no NaN or infinity values;
- explicit JSON `null` where permitted; and
- one documented canonicalization-version identifier.

Provide one shared serialization function used by both publication and validation.

The manifest digest must be:

```text
artifact_manifest_sha256 =
    sha256(canonical_json(artifact_manifest))
```

The manifest must not contain its own checksum.

## Completion marker

`integrity/publication_complete.json` must be logged last.

It must contain:

- completion-marker schema version;
- model name;
- exact numeric model version;
- `model_version_id`;
- MLflow run ID;
- `pipeline_sha256`;
- `artifact_manifest_sha256`;
- completion timestamp in UTC; and
- publisher version.

The completion marker is not part of the manifest digest.

A registered version is integrity-complete only when:

- the completion marker exists;
- the run ID matches the registered model’s source run;
- the model name and numeric version match;
- the `model_version_id` matches;
- the manifest exists and passes its JSON Schema;
- the manifest checksum matches the completion marker;
- every protected artifact exists;
- every protected artifact’s size and SHA-256 digest match;
- the pipeline checksum matches;
- reference purposes and privacy metadata remain valid;
- the fitted pipeline can be safely loaded after checksum verification; and
- the smoke prediction succeeds.

A registered version created before publication completes may remain in DagsHub, but it must be marked incomplete and must not be deployment-eligible.

## Publishing flow

Extend the existing DagsHub tracking and registration flow:

1. Complete local training and evaluation.
2. Validate privacy and contract requirements.
3. Open one DagsHub MLflow run.
4. Log parameters, metrics, contracts, references, and approved artifacts.
5. Log the exact fitted MLflow model.
6. Register the model under `churn_predictor`.
7. Resolve the exact numeric registered version.
8. Download or enumerate the final logged artifact set.
9. Calculate protected artifact sizes and SHA-256 digests.
10. Generate and validate the canonical manifest.
11. Log `integrity/artifact_manifest.json`.
12. Calculate `artifact_manifest_sha256`.
13. Set integrity tags on the MLflow run and registered model version.
14. Independently validate the completed publication.
15. Log `integrity/publication_complete.json` last.
16. Mark the model version integrity status as complete.

Suggested model-version tags:

```text
integrity_status=complete
artifact_manifest_schema_version=1.0.0
artifact_manifest_sha256=<digest>
pipeline_sha256=<digest>
source_commit_sha=<commit>
```

If any step fails:

- return a non-zero status when registration was required;
- do not publish the completion marker;
- set `integrity_status=incomplete` when possible;
- do not treat the version as deployment-eligible; and
- do not silently fall back to another version.

## Validation command

Extend the existing command:

```bash
python -m src.mlops validate-model \
  --model-uri models:/churn_predictor/<numeric-version> \
  --output json
```

Do not add a separate `src.artifacts` CLI.

For integrity-enabled model versions, validation must additionally:

- download the manifest and completion marker;
- validate the manifest against its versioned JSON Schema;
- recalculate canonical manifest serialization;
- recalculate `artifact_manifest_sha256`;
- verify the completion marker;
- verify run and registered-version linkage;
- verify every protected file’s size and SHA-256 digest;
- verify the complete MLflow model tree;
- verify pipeline checksum consistency;
- verify contract-version consistency;
- verify drift and evaluation reference purposes;
- reject prohibited reference columns;
- load the pipeline only after integrity checks pass;
- run a schema-valid smoke prediction; and
- return machine-readable integrity results.

Example output:

```json
{
  "valid": true,
  "integrity_status": "complete",
  "model_name": "churn_predictor",
  "model_version": "4",
  "model_version_id": "dagshub:owner/repository:churn_predictor:4",
  "mlflow_run_id": "<run-id>",
  "pipeline_sha256": "<digest>",
  "artifact_manifest_sha256": "<digest>",
  "manifest_schema_version": "1.0.0"
}
```

Validation must return a non-zero exit code when any protected artifact or identity check fails.

## Backward compatibility

Registered versions created before this feature must not be silently presented as integrity-verified.

For legacy versions, `validate-model` must clearly report:

```json
{
  "integrity_status": "legacy"
}
```

Legacy handling must be explicit and documented.

After the first integrity-enabled production version is registered, new production deployments must require:

```text
integrity_status=complete
```

Do not retrofit or mutate historical registered versions merely to make them appear integrity-enabled.

## Modal deployment integration

Extend the deployment contract with:

```text
EXPECTED_ARTIFACT_MANIFEST_SHA256
```

`prepare-deployment` must:

- require the expected manifest checksum for integrity-enabled versions;
- validate the exact registered version;
- verify the manifest and completion marker;
- verify every protected artifact before loading the model;
- confirm the expected manifest checksum;
- include the verified checksum in `deployment_metadata.json`; and
- continue copying only inference-required files into `build/model`.

The Modal image must not contain drift references, evaluation references, training metrics, or unrestricted DagsHub artifacts.

Modal startup must verify:

- exact numeric model version;
- expected run ID;
- expected `model_version_id`;
- expected pipeline checksum;
- expected artifact-manifest checksum;
- feature-schema compatibility;
- model deserialization;
- smoke prediction; and
- database connectivity.

Startup must fail before readiness when any expected identity or checksum differs.

## Runtime observability

`GET /health` must include:

```json
{
  "artifact_manifest_sha256": "<verified-digest>",
  "integrity_status": "complete"
}
```

Safe startup and prediction logs must include:

- deployment ID;
- numeric model version;
- MLflow run ID;
- `model_version_id`;
- pipeline checksum;
- artifact-manifest checksum; and
- integrity status.

Logs must not include:

- input feature values;
- raw reference rows;
- database URLs;
- DagsHub tokens;
- authentication headers; or
- raw customer identifiers.

The existing single and batch response identity fields do not need to be redesigned.

## Security requirements

- Calculate and verify checksums before deserializing the fitted pipeline.
- Treat the serialized sklearn pipeline as a trusted executable artifact.
- Accept artifacts only from the configured DagsHub repository.
- Never include credentials in manifests or completion markers.
- Never include raw customer identifiers.
- Never include absolute local filesystem paths.
- Keep DagsHub publishing credentials outside Modal.
- Keep the Neon URL only in the Modal runtime secret.
- Redact secrets from validation and publication failures.
- Log checksums and artifact identities, not artifact contents.

## Tests

Add focused coverage for:

### Manifest generation

- deterministic canonical JSON serialization;
- stable protected-artifact ordering;
- SHA-256 and byte-size calculation;
- repository-independent path normalization;
- JSON Schema validation;
- rejection of absolute paths;
- rejection of duplicate artifact paths;
- rejection of secrets or prohibited fields in manifest metadata.

### Publication integrity

- complete manifest publication;
- completion marker written last;
- missing protected artifact rejection;
- changed artifact checksum rejection;
- changed artifact size rejection;
- manifest checksum mismatch rejection;
- run-ID mismatch rejection;
- numeric model-version mismatch rejection;
- `model_version_id` mismatch rejection;
- pipeline checksum mismatch rejection;
- incomplete publication is not deployment-eligible.

### Existing validator

- valid integrity-enabled version passes;
- legacy version reports `integrity_status=legacy`;
- malformed manifest fails;
- missing completion marker fails;
- wrong reference purpose fails;
- prohibited reference columns fail;
- pipeline loads only after integrity verification;
- smoke prediction failure returns non-zero;
- JSON output contains no credentials.

### Deployment

- `prepare-deployment` accepts the correct expected manifest checksum;
- `prepare-deployment` rejects a mismatched manifest checksum;
- deployment metadata includes the verified manifest checksum;
- startup rejects missing integrity metadata;
- startup rejects mismatched integrity metadata;
- health reports verified integrity status and manifest checksum;
- inference package excludes training and reference artifacts.

External DagsHub integration tests must remain opt-in.

## Documentation

Document:

- why DagsHub MLflow remains the only model registry;
- the difference between numeric model version, `model_version_id`, pipeline checksum, and artifact-manifest checksum;
- protected and excluded artifact classes;
- manifest canonicalization rules;
- publication ordering;
- incomplete-publication behavior;
- legacy-version behavior;
- how to validate an exact registered version;
- how to prepare an integrity-verified Modal package;
- how to configure `EXPECTED_ARTIFACT_MANIFEST_SHA256`;
- how to inspect integrity metadata through `/health`;
- how to redeploy a historical integrity-enabled numeric version; and
- credential and trust boundaries.

## Acceptance criteria

- [ ] DagsHub MLflow remains the only tracking backend and model registry.
- [ ] No custom historical `models/` store or content-hash model identity is introduced.
- [ ] Every newly completed production registration has a versioned artifact manifest.
- [ ] Every protected artifact has a SHA-256 digest and byte size.
- [ ] Manifest serialization is canonical and independently reproducible.
- [ ] The manifest checksum is recorded on both the MLflow run and model version.
- [ ] A completion marker is published only after full validation succeeds.
- [ ] Incomplete versions cannot be packaged for production.
- [ ] Existing `validate-model` independently verifies the complete artifact set.
- [ ] Legacy model versions are explicitly reported as legacy.
- [ ] `prepare-deployment` requires the expected manifest checksum for integrity-enabled versions.
- [ ] Modal remains pinned to one exact numeric registered version.
- [ ] The Modal image remains inference-only.
- [ ] Startup rejects a missing, corrupt, incomplete, or incorrectly pinned publication.
- [ ] Health metadata exposes the verified manifest checksum and integrity status.
- [ ] Safe operational logs include integrity identity without customer data or credentials.
- [ ] Unit and opt-in integration tests cover publication, validation, deployment, and failure cases.
- [ ] Publishing, validation, deployment, legacy handling, and recovery are documented.

## Out of scope

- Replacing DagsHub MLflow with a custom registry.
- Content-addressed model-version IDs.
- Local or bucket-backed historical model stores.
- Runtime model hot-swapping.
- Automatic promotion or aliases.
- Champion/challenger deployment.
- Automatic retraining.
- Drift report generation.
- Outcome ingestion.
- Prediction persistence.
- Operational monitoring tables.
- Changes to the approved feature, privacy, identity, or monitoring contracts.
```

This revised issue preserves the useful missing capability—whole-publication integrity—without recreating the training, registry, reference, packaging, deployment, and identity work already completed.
